library/kernel_settings.py

```diff
@@ -0,0 +1,990 @@
+      NOTE: the options list may be incomplete - the actual options are generated
+      dynamically from tuned, for the current options supported by the version
+      of tuned, which are the tuned supported plugins.  Only the most common
+      options are listed.  See the tuned documentation for the full list and
+      more information.
+
+options:
+    sysctl:
+        description:
+            - key/value pairs of sysctl settings to apply - this works mostly
+              like the `sysctl` module except that `/etc/sysctl.conf` and files
+              under `/etc/sysctl.d` are not used.
+        required: false
+    sysfs:
+        description:
+            - key/value pairs of sysfs settings to apply
+        required: false
+    bootloader:
+        description:
+            - the `cmdline` option can be used to set, add, or delete
+              kernel command line options.  See EXAMPLES for some examples
+              of how to use this option.
+              Note that this uses the tuned implementation, which adds these
+              options to whatever the default bootloader command line arguments
+              tuned is historically used to add/delete performance related
+              kernel command line arguments e.g. `spectre_v2=off`.  If you need
+              more general purpose bootloader configuration, you should use
+              a bootloader module/role.
+    operations:
+        description:
+            - This a dict.  Each key is the name of a section to operate on -
+              `sysctl`, `sysfs`, etc. same as the module setting options.  The
+              value is a list of operations to apply.  Each individual operation
+              is a dict.  Mostly this is used to delete values from sections i.e.
+              to restore their default values.  You can delete a value, then add
+              it back, to avoid having the role issue a `Conflict` warning that a
+              value conflicts with an existing value.  See `EXAMPLES` below.
+    purge:
+        description:
+            - Remove the current kernel_settings, whatever they are, and force
+              the given settings to be the current and only settings
+
+author:
+    - Rich Megginson (rmeggins@redhat.com)
+"""
+
+EXAMPLES = """
+# add the sysctl vm.max_mmap_regions, and disable spectre/meltdown security
+kernel_settings:
+    sysctl:
+        vm.max_mmap_regions: 262144
+    sysfs:
+        /sys/kernel/debug/x86/pti_enabled: 0
+        /sys/kernel/debug/x86/retp_enabled: 0
+        /sys/kernel/debug/x86/ibrs_enabled: 0
+
+# replace the existing sysctl section with the specified section
+# delete the /sys/kernel/debug/x86/retp_enabled setting
+# completely remove the vm section
+# add the bootloader cmdline arguments spectre_v2=off nopti
+# remove the bootloader cmdline arguments quiet splash
+kernel_settings:
+    sysctl:
+        vm.max_mmap_regions: 262144
+    bootloader:
+        cmdline: spectre_v2=off nopti
+    operations:
```

Comment:
This (and also the `DOCUMENTATION` variable above) suggests that the `operations` are on the same level as `sysctl` etc. Whereas the README suggests that the `operations` are on the same level as `kernel_settings_parameters`. Is there an inconsistency in the docs, or am I misunderstanding something? (I'm not really familiar with ansible, so that's a real possibility.)

Reply:
This documentation is for the module, not the role.  The module is not intended to be used outside of the role.  I guess you could say the role is the "interface" and the module is the default "implementation" of the role interface.